src/OrchardCore/OrchardCore.ResourceManagement/FileVersionProvider.cs

```diff
@@ -0,0 +1,146 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Antiforgery.Internal;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.FileProviders;
+using OrchardCore.Abstractions.Modules.FileProviders;
+
+namespace OrchardCore.ResourceManagement
+{
+    /// <summary>
+    /// Provides version hash for a specified file.
+    /// </summary>
+    public class FileVersionProvider : IFileVersionProvider
+    {
+        private const string VersionKey = "v";
+        private static readonly char[] QueryStringAndFragmentTokens = new[] { '?', '#' };
+
+        private readonly IEnumerable<IFileProvider> _registeredFileProviders;
+        private readonly IFileProvider _webRootfileProvider;
+
+        private IList<IFileProvider> _combinedFileProviders;
+        private readonly IMemoryCache _cache;
+        public FileVersionProvider(
+            IEnumerable<IFileProvider> registeredFileProviders,
+            IHostingEnvironment environment,
+            IMemoryCache cache
+            )
+        {
+            _registeredFileProviders = registeredFileProviders;
+            _webRootfileProvider = environment.WebRootFileProvider;
+            _cache = cache;
+        }
+
+        public string AddFileVersionToPath(PathString requestPathBase, string path)
+        {
+            if (path == null)
+            {
+                throw new ArgumentNullException(nameof(path));
+            }
+
+            // Build _combinedFileProviders if it has not been built
+            if (_combinedFileProviders == null)
```

Comment:
So i removed this part.

Note: Because this is a tenant singleton, before this part could be called concurrently.